### PR TITLE
SDK-113 Verify JDK version on setup

### PR DESCRIPTION
**Pull request by tmarzeion**

### CODE REVIEW

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java [hunk 1]
```diff
@@ -191,10 +191,20 @@ else if(isCreatePlatform) {
...
         }
 
-        wizard.promptForJavaHomeIfMissing(server);
 
         server.setUnspecifiedToDefault();
         configureVersion(server, isCreatePlatform);
+
+        String platformVersion = server.getPlatformVersion();
+        if (platformVersion.charAt(0) == 1) {
+            wizard.showMessage("Note: JDK 1.7 is needed for platform version " + platformVersion + ".");
```
Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java]: @rkorytkowski What about earlier versions of JDK? Should v1.* platform work on JDK v1.6?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java [hunk 2]
```diff
@@ -305,25 +305,70 @@ public void promptForJavaHomeIfMissing(Server server) {
...
+                    promptForJavaHomeIfMissing(server);
+                }
             }
         }
+    }
 
-
+    private String extractJavaVersionFromPath(String path) {
+        path = path.substring(0, path.length()-4);
+        File jdkReleasePropertiesFile = new File(path, "release");
```
Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Thats how JDK version is extracted from JDK directory.
Works for Oracle releases.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java (hunk 1 above)
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java]: Nope, we dropped support for 1.6.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java (hunk 2 above)
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: That's not correct. You should execute 'java -version' using processbuilder to get the version.

Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Actually sorry, I should have asked if your solution works for OpenJDK? If yes, then it's correct.

Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Fixed :)

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java [hunk 3]
```diff
@@ -305,25 +307,74 @@ public void promptForJavaHomeIfMissing(Server server) {
...
+
+        String result = null;
 
+        try {
+            final Process process = processBuilder.start();
+            BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+
+            result = stdInput.readLine();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Do not print stack trace.... Rethrow as RuntimeException.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java [hunk 4]
```diff
@@ -305,25 +307,74 @@ public void promptForJavaHomeIfMissing(Server server) {
...
+
+        ProcessBuilder processBuilder = new ProcessBuilder(commands);
+
+        processBuilder.directory(new File(path + File.separator + "bin"));
+
+        String result = null;
 
+        try {
+            final Process process = processBuilder.start();
+            BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getErrorStream()));
```
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Why getErrorStream?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java [hunk 5]
```diff
@@ -305,25 +307,74 @@ public void promptForJavaHomeIfMissing(Server server) {
...
+
+            result = stdInput.readLine();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
 
+        return result.split("\"")[1];
```
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Could you please use some regex instead? It seems error prone if the ouput changes even slightly.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java (hunk 4 above)
Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Fixed by using `processBuilder.redirectErrorStream(true)`

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java (hunk 5 above)
Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Fixed

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java (hunk 3 above)
Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Sorry, I've probably missed that.
Fixed

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java [hunk 6]
```diff
@@ -317,19 +321,78 @@ public void promptForJavaHomeIfMissing(Server server) {
...
+                    server.setJavaHome(null);
+                }
+                else if (System.getProperty("java.version").startsWith(requiredJdkVersion)) {
+                    server.setJavaHome(null);
+                }
+                else {
+                    showMessage("Your JAVA_HOME version doesn't fit platform requirements:");
+                    showMessage("JAVA_HOME version: " + System.getProperty("java.version"));
+                    showMessage("Required: " + requiredJdkVersion);
+                    promptForJavaHomeIfMissing(server);
```
Comment (rkorytkowski) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Let's make it less strict and require only java 8 for platform 2.x, but allow using java 8 for platform 1.x with additional confirmation from the user "It is not recommended to run OpenMRS platform {versionHere} on JDK 8. Would you like to select the recommended JDK 7 instead? [Y/n]:"

Comment (tmarzeion) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/DefaultWizard.java]: Fixed

**Pull request merged**